[16.0][IMP] l10n_be_cooperator: tax shelter report

victor-champonnois opened the pull request:
Makes tax shelter attestation consistent with administration templates

internal task : https://gestion.coopiteasy.be/web#id=11176&action=475&active_id=524&model=project.task&view_type=form&menu_id=536

@huguesdk

victor-champonnois commented:
@huguesdk not ready yet : I need to fix the tests

victor-champonnois commented:
@huguesdk now ready for review.
Note : I was first tempted to remove the table with a recap of the shares in the certificates, since it is not legally required, and would greatly simplify the module.
But I wasn't sure if we wanted to do that, since this functionality may still be interesting. So I reverted the commit. 
I leave you to decide :)
If we do it, I would have to remove some tests.

huguesdk commented on l10n_be_cooperator/models/tax_shelter_declaration.py:
```diff
@@ -17,22 +17,23 @@ class TaxShelterDeclaration(models.Model):
-    date_to = fields.Date(string="Date to", required=True)
-    month_from = fields.Char(string="Month from", required=True)
-    month_to = fields.Char(string="Month to", required=True)
-    tax_shelter_percentage = fields.Selection(
-        [("25", "25%"), ("30", "30%"), ("45", "45%")],
-        string="Tax Shelter percentage",
```
i think that it would be good to keep the `tax_shelter_percentage` field, but as a stored computed read-only field (and simple text instead of selection). this would be useful for 2 things:

1.  have the logic to compute the percentage from the type defined only once instead of being duplicated in the email template and in the report.
2.  make the migration script easier: it could simply be a post script that fills in the `tax_shelter_report_type` corresponding to the percentage.

victor-champonnois replied:
I agree -> Done

huguesdk commented on l10n_be_cooperator/models/tax_shelter_declaration.py:
```diff
@@ -17,22 +17,23 @@ class TaxShelterDeclaration(models.Model):
     _description = "Tax Shelter Declaration"
 
     name = fields.Char(string="Declaration year", required=True)
-    fiscal_year = fields.Char(string="Fiscal year", required=True)
```
i agree with the changes that remove the `string` when it corresponds to the field name, but keep in mind that this will capitalize each word. this is what we want (standard odoo practice), but the translations need to be updated.

victor-champonnois replied:
OK, the translations should be done after the merge with weblate I think.

huguesdk replied:
as discussed, it is better that minor changes (capitalization, language mistakes) are dealt with by the developer directly, to avoid useless work for translators.

huguesdk commented on l10n_be_cooperator/README.rst:
your first commit refers to the official documents. i think that it would be nice to add these links also to the readme.

victor-champonnois replied:
done

huguesdk commented on l10n_be_cooperator/models/tax_shelter_declaration.py:
```diff
@@ -17,22 +17,23 @@ class TaxShelterDeclaration(models.Model):
-        string="Tax Shelter percentage",
+    date_from = fields.Date(required=True)
+    date_to = fields.Date(required=True)
+    tax_shelter_report_type = fields.Selection(
```
i think that a more appropriate name would be `tax_shelter_type`, because this is not related to the report.

victor-champonnois replied:
done

huguesdk commented on l10n_be_cooperator/models/tax_shelter_declaration.py:
```diff
@@ -17,22 +17,23 @@ class TaxShelterDeclaration(models.Model):
-        string="Tax Shelter percentage",
+    date_from = fields.Date(required=True)
+    date_to = fields.Date(required=True)
+    tax_shelter_report_type = fields.Selection(
```
a migration script is needed to set the value of this field.

huguesdk commented on l10n_be_cooperator/report/tax_shelter_report_templates.xml:
```diff
@@ -104,47 +105,69 @@ SPDX-License-Identifier: AGPL-3.0-or-later
+                        <span
+                            t-if="report_type=='start_up_small' or report_type=='start_up_micro'"
+                        >14526</span>
+                        <span t-if="report_type=='scale_up'">14527</span>
+                        , CIR 92 sur les revenus
```
the comma should be just after `</span>`, otherwise a (hideous) space appears before the comma.

victor-champonnois replied:
fixed (indeed, what's more hideous than a misplaced space...^^)

huguesdk commented on l10n_be_cooperator/report/tax_shelter_report_templates.xml:
```diff
@@ -104,47 +105,69 @@ SPDX-License-Identifier: AGPL-3.0-or-later
                     </h2>
+                    <em>
 
+                        <span
+                            t-if="report_type=='start_up_small' or report_type=='start_up_micro'"
+                        >
+                            Attestation délivrée dans le cadre de la réduction d’impôt pour l’acquisition de nouvelles
+                        actions ou parts d’entreprises qui débutent prévue à l’article 14526, CIR 92 et à l’article
+                        6312/1, AR/CIR 92.
+                    </span>
+
+                    <span t-if="report_type=='scale_up'">
+                        Attestation délivrée dans le cadre de la réduction d'impôt pour l'acquisition de
+                        nouvelles actions ou parts d'entreprises en croissance prévue à l'article 14527, CIR 92 et
+                        à l'article 6312/3, AR/CIR 92.
+                    </span>
```
some cosmetic changes to make here:

1.  the first paragraph uses the typographic apostrophe (’, aka “right single quotation mark”), while the regular apostrophe (') is used elsewhere. it should be uniform.
2.  the 3 first lines of the first paragraph are too indented compared to the rest.

victor-champonnois replied:
fixed, and ran an automatic xml formatter to avoid other inconsistencies

victor-champonnois replied:
actually, no the automatic formatter makes a mess, just corrected what you suggest

huguesdk commented on l10n_be_cooperator/report/tax_shelter_report_templates.xml:
```diff
@@ -104,47 +105,69 @@ SPDX-License-Identifier: AGPL-3.0-or-later
+                                <span
+                                    t-if="report_type=='start_up_small' or report_type=='start_up_micro'"
+                                >
+                                    les conditions prévues aux § 1er et 2 de l'article 14527, CIR 92, sont respectées
+                                </span>
+                                <span t-if="report_type=='scale_up'">
+                                    les conditions prévues au § 3, alinéa 1er, de l’article 14526, CIR 92, sont respectées
```
aren’t the article numbers swapped here?

victor-champonnois replied:
yes, nice catch !

huguesdk commented on l10n_be_cooperator/report/tax_shelter_report_templates.xml:
```diff
@@ -104,47 +105,69 @@ SPDX-License-Identifier: AGPL-3.0-or-later
+                            t-if="report_type=='start_up_small' or report_type=='start_up_micro'"
+                        >
+                            Attestation délivrée dans le cadre de la réduction d’impôt pour l’acquisition de nouvelles
+                        actions ou parts d’entreprises qui débutent prévue à l’article 14526, CIR 92 et à l’article
```
the official name of the article is 145<sup>26</sup> (`145<sup>26</sup>`). same for 145<sup>27</sup> and other occurrences in this module.

victor-champonnois replied:
fixed

huguesdk commented on l10n_be_cooperator/report/tax_shelter_report_templates.xml:
```diff
@@ -104,47 +105,69 @@ SPDX-License-Identifier: AGPL-3.0-or-later
+                    <span t-if="report_type=='scale_up'">
+                        Attestation délivrée dans le cadre de la réduction d'impôt pour l'acquisition de
+                        nouvelles actions ou parts d'entreprises en croissance prévue à l'article 14527, CIR 92 et
+                        à l'article 6312/3, AR/CIR 92.
```
the official name of the article is 63<sup>12/3</sup> (`63<sup>12/3</sup>`).

victor-champonnois replied:
fixed

victor-champonnois commented:
@huguesdk thank you for the extensive review!
- I added your requests
- I removed the commits that removed the table, since the table is necessary and useful
- I added a commit to remove the "capital before" and "capital after". These columns are not legally necessary, and some companies might not like it.  
- made the table format less weird (according to my own personal taste)

huguesdk commented on l10n_be_cooperator/migrations/16.0.1.1.0/pre-migration.py:
```diff
@@ -0,0 +1,40 @@
+
+@openupgrade.migrate()
+def migrate(env, version):
+    if not openupgrade.column_exists(
+        env.cr, "tax_shelter_declaration", "tax_shelter_type"
+    ):
```
this test is not needed, and in fact, the dependency on `openupgradelib` is not necessary. a normal odoo migration script should do.

victor-champonnois replied:
removed the test. I think I still need openupgrade for `openupgrade.logged_query`, it doesn't seem to work otherwise

huguesdk replied:
if you declare the migrate function (without the openupgrade decorator) as:
```python
def migrate(cr, version):
```
you can do:
```python
cr.execute("some query")
```

huguesdk commented on l10n_be_cooperator/models/tax_shelter_declaration.py:
```diff
@@ -17,22 +17,28 @@ class TaxShelterDeclaration(models.Model):
+        [
+            ("scale_up", _("25% - Scale-up")),
+            ("start_up_small", _("30% - Start-up (petite société)")),
+            ("start_up_micro", _("45% - Start-up (micro société)")),
```
as all the strings coming from the fields are in english, i think it should be more logical that these are also: “small company”, “micro company”.

huguesdk replied:
also, calling `_()` is unnecessary: this is automatic for selection fields.

huguesdk replied:
resolved by my next commit (in which `_()` is still needed).

huguesdk commented on l10n_be_cooperator/models/tax_shelter_declaration.py:
```diff
@@ -17,22 +17,28 @@ class TaxShelterDeclaration(models.Model):
-        string="Tax Shelter percentage",
-        required=True,
+        compute="_compute_tax_shelter_percentage",
+        readonly=True,
```
great! i first thought that it would be better to have it as a simple string and stored, but i changed my mind and it is good like that:

*   no need to store it, at it is derived from the type, and there is a pre-migration script to set the values
*   having the labels directly in the field allows to handle the localization (spacing, using the percentage sign or not,…) once in the translation files instead of handling each case separately in the email templates and the reports.

`readonly=True` should be removed, as it is the default for computed fields.

huguesdk replied:
i eventually changed it to float, to avoid repetition of the different cases while ensuring correct localization in case there are decimals.

victor-champonnois replied:
If it is not stored, I get the warning `tax.shelter.declaration: inconsistent 'compute_sudo' for computed fields: tax_shelter_percentage, tax_shelter_capital_limit ` , So I've put the `store` back

huguesdk commented on l10n_be_cooperator/migrations/16.0.1.1.0/pre-migration.py:
```diff
@@ -0,0 +1,40 @@
+            """
+            UPDATE tax_shelter_declaration
+            SET tax_shelter_type = 'scale_up'
+            WHERE tax_shelter_declaration.tax_shelter_percentage = '25'
```
one update statement is enough:

```suggestion
            UPDATE tax_shelter_declaration
            SET tax_shelter_type = CASE
                WHEN tax_shelter_percentage = '25' THEN 'scale_up'
                WHEN tax_shelter_percentage = '30' THEN 'start_up_small'
                WHEN tax_shelter_percentage = '45' THEN 'start_up_micro'
            END
```

huguesdk commented on l10n_be_cooperator/report/tax_shelter_report_templates.xml:
```diff
@@ -12,32 +12,20 @@ SPDX-License-Identifier: AGPL-3.0-or-later
                     <span t-field="line.transaction_date" />
                 </td>
-                <td class="text-end">
+                <td class="text-start">
```
these were aligned at the end to ensure decimals line up in case there are multiple lines. but since there is no total, i think it’s ok like this, and it looks better, indeed.

huguesdk reviewed (changes requested):
thanks, we’re almost there!

victor-champonnois commented:
@huguesdk 
- updated the po and pot files
- updated the migration script (but kept the openupgrade dependency)
- added store on tax shelter percentages because of a warning

victor-champonnois commented:
@huguesdk done (although I don't see the query in the log, which is a bit sad)

victor-champonnois commented:
@huguesdk can we merge this PR ? It would be nice to to have it ready for the next tax shelter period

huguesdk commented:
@victor-champonnois i checked the text of the certificates and noticed that the list of requirements did not change according to the report type and the tax shelter type. i fixed this and did also some improvements in the report (change some wording, use `t` tags instead of `span` tags for conditions, improve indentation). i also added and improved some translations. please test on your side and tell me if you think it is ready.

edit: if ready, in addition to squashing commits, adding news fragments would be nice.

victor-champonnois commented:
@huguesdk thanks for the change, I've tested and it looks OK. I've squashed all the commits, keeping the detail of what has been done in the commit message.

victor-champonnois commented:
and I added a small newsfragment. Ready to merge !

huguesdk reviewed (approved):
🎉 many thanks, @victor-champonnois!

huguesdk commented:
/ocabot merge nobump
